File: java/lang/src/main/java/otterop/lang/WrapperOOPIterator.java
package otterop.lang;

import java.util.Iterator;
import java.util.function.Function;

public class WrapperOOPIterator<FROM,TO> implements OOPIterator<TO>, Iterator<TO> {
    private Iterator<FROM> it;
    private Function<FROM,TO> wrap;

    public WrapperOOPIterator(Iterator<FROM> it, Function<FROM,TO> wrap) {
        this.it = it;
        this.wrap = wrap;
    }

    @Override
    public boolean hasNext() {
        return this.it.hasNext();
    }

    @Override
    public TO next() {
        return this.wrap.apply(it.next());
    }
}

File: python/lang/otterop/lang/wrapper_oop_iterator.py
class WrapperOOPIterator:

    def __init__(self, it, convert):
        self._it = it
        self._convert = convert
        self._next = None
        self._has_next = None

    def __next__(self):
        return self.next()

    def has_next(self):
        if self._has_next is None:
            try:
                self._next = self._it.__next__()
                if self._convert is not None:
                    self._next = self._convert(self._next)
                self._has_next = True
            except StopIteration as e:
                self._has_next = False
        return self._has_next

    def next(self):
        if self._has_next is None:
            self.has_next()
        if self._has_next is False:
            raise StopIteration()
        self._has_next = None
        return self._next

